### PR TITLE
Fix #269: Browse original type trees in Dependency

### LINKED ISSUE #269: Zinc is undercompiling in case of type projection where Scalac expands to another type
In case included in #270 zinc is undercompiling.

I don't get to the bottom of problem (yet) but I think this is caused by the way how we handle type fields.

### PR DESCRIPTION
**Pull request by jvican**
From the commit message:

```
Fix #269: Traverse original trees in `Dependency`

Original type trees have to be traversed if present to correctly handle
dependencies specified in explicit, user-defined types that are expanded
by the compiler, in cases such as type projections.

The example `fuzzy-types` is undercompiling because the prefix
`FactoryProvider` is totally lost at `Usage` after typer. This is one of
these cases where Scalac is expanding types and not leaving any trace to
keep track of the original dependency.

`FactoryProvider.type#MyFactory#Product` becomes `foo.B with
foo.Nil#Product`, and therefore the type dependency traverser doesn't
see `FactoryProvider`. Traversing original type trees fixes the issue.
```

### CODE REVIEW

**Comment by jvican** [comment]
The first two commits should go away as soon as #270 is merged.

**Comment by jvican** [comment]
This has been rebased, it's awaiting for review.

**Review by eed3si9n — APPROVED** [review]
LGTM

File: internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala
Comment (Duhemm) [inline internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala:387]: Can we get cycles here?

`inspectedOriginalTrees` was introduced because there were some macros that had themselves as original trees.

Comment (jvican) [reply internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala:387]: *Theoretically* yes.
  
In addition to that, it serves us as a caching layer, which was my primary goal when implementing it.

**Comment by jvican** [comment]
Can you have a look at this @dwijnand?
  
Also, FYI @eed3si9n & @dwijnand, I was thinking it would be a good idea to cut a release of Zinc soon that includes:
  
1. This fix.
2. Handling of private members (working on it).
3. Handling of constants (working on it).
  
WDYT guys?

**Comment by romanowski** [comment]
@jvican Did you run benchmarks? I wonder how much overhead is added here.

I think release is good idea and as you may expect I vote for my pr #267 be included :)

**Comment by jvican** [comment]
I have not run benchmarks on this one. Why? Because:
  
1. It's a correctness issue -- it fixes a bug.
2. We're already doing the original tree traversal in `ExtractUsedNames`, this PR only does the equivalent in `Dependency`. E.g. we're not introducing new behaviour.
  
We haven't discussed this publicly, but I think we should not run benchmarks for bugfixes, only for new features and PRs that change the structure of the algorithm (like #267, for instance, that happens to do both).
  
Knowing the benchmark results for every bugfix will take the author of the PR an insanely amount of time and will not give anything to the maintainers: it's a bugfix, they have to merge it in any case.
  
I would be for checking performance results on bugfixes that touch a lot of code and sensible parts of the Scalac phases. But as this PR only propagates a behaviour we are doing in `ExtractUsedNames`, I suggest not to run them.

**Comment by eed3si9n** [comment]
@jvican I created GitHub milestone named 1.0.0-X11 to track the next release - https://github.com/sbt/zinc/milestone/2

**Comment by jvican** [comment]
Thanks @eed3si9n!

**Review by dwijnand — APPROVED** [review]

**Pull request merged**
